syntactic-whitelist/src/test/scala/TestRunner.scala

```diff
@@ -0,0 +1,245 @@
+    val actualRes = checker.checkFile(s"$TEST_RES_DIR/$filename.$TEST_FILES_EXT")
+    matcher(actualRes)
+  }
+
+}
+
+object TestRunner {
+
+  private val uidGenerator = new AtomicInteger(0)
+
+  private val DEFAULT_DIALECT = dialects.Scala3
+  private val TEST_RES_DIR = "src/test/res"
+  private val TEST_FILES_EXT = "scala"
+
+  // index of the frame of the StackWalker that corresponds to the test name, assuming that StackWalker was
+  // instantiated in createTest
+  private val TEST_METHOD_FRAME_IDX = 1
+
+  /**
+   * @return a builder for TestRunner
+   * @param testController the controller that will be used to ensure that the test is actually run <p>
+   * By default it will use Scala3 as a dialect, but this can be overriden <p>
+   * <strong> This method should be called at top level inside the tests, it should not be nested
+   * (for error reporting to work correctly) </strong>
+   */
+  def createTest(testController: TestController): Builder = {
+    val testName = StackWalker.getInstance().walk(frames => {
+      val frame = frames.limit(TEST_METHOD_FRAME_IDX+1).toList.get(TEST_METHOD_FRAME_IDX)
+      frame.getMethodName
+    })
+    new Builder(testController, testName)
+  }
+
+  // builder for TestRunner
+  class Builder protected[TestRunner](testController: TestController, testName: String) {
+    private var matcher: Option[CheckResult => Unit] = None
+    private var filename: Option[String] = None
+    private val features: ListBuffer[Feature] = ListBuffer.empty
+    private var dialect: Option[Dialect] = None
+    private val uid = uidGenerator.incrementAndGet()
+
+    {
+      // tell the controller that a test with this uid must be executed
+      testController.registerTest(uid, testName)
+    }
+
+    /**
+     * @param _filename name of the file to run the checker on (name only, without directory or extension)
+     */
+    def onFile(_filename: String): Builder = {
+      if (this.filename.isDefined){
+        throw new IllegalStateException("filename set more than once")
+      }
+      this.filename = Some(_filename)
+      this
+    }
+
+    /**
+     * @param _dialect the Scala dialect to use
+     */
+    def withDialect(_dialect: Dialect): Builder = {
+      if (this.dialect.isDefined){
+        throw new IllegalStateException("dialect is set more than once")
+      }
+      this.dialect = Some(_dialect)
+      this
+    }
+
+    /**
+     * @param _features features that the Checker should allow
+     */
+    def withFeatures(_features: Feature*): Builder = {
+      this.features.addAll(_features)
+      this
+    }
+
+    /**
+     * @param _features features that the Checker should allow
+     */
+    def withFeatures(_features: List[Feature]): Builder = {
+      this.features.addAll(_features)
+      this
+    }
+
+    /**
+     * @param _features features that the Checker should allow
+     */
+    def exceptFeatures(_features: Feature*): Builder = {
+      this.features.filterInPlace(!_features.contains(_))
```

Comment:
I did not think of that. What should I do then? Only allow `AtomicFeature`s in `_features`? Like that:
```Scala
def exceptFeatures(_features: AtomicFeature*): Builder
```

Reply:
I propose the following:
- instances of `AtomicFeature` are defined by us only. This ensures we cover the whole language (so, it gives support for blacklisting language features), and this ensures that atomic features can only be referenced by their public identifier (which is a static object),
- make `CompositeFeature` a case class, or override the equality definition to be structural.

The drawback of this design would be that users would not be able to define their own atomic features.

Reply:
So what should I do here? I think that preventing the user from defining their own `AtomicFeature`s is very restrictive. And if we want to make equality structural on `CompositeFeature` we will have a problem with the fact that `CompositeFeature` has a list as an attribute, so the order of the features matters.
Is it sufficient if I simply use a set instead of the list and override equals/hashCode to make equality structural on `CompositeFeature`, without preventing the user from defining custom `AtomicFeature`s?

Reply:
> Is it sufficient if I simply use a set instead of the list

Yes, that’s a good point.

>  and override equals/hashCode to make equality structural on `CompositeFeature`,
> without preventing the user from defining custom `AtomicFeature`s?

That will be enough as long as users use the predefined `AtomicFeature`s, but if they define their features, then they will have to also override hashCode/equals.

Reply:
OK sorry, I was focused on equality for `CompositeFeature` and I had not thought of the problem with equality in `AtomicFeature`. So I did what you said and forbid customs `AtomicFeature`s (ed1bdae323805d800895e8232a28a98e1dff9c4b), thank you for your help!